### PR TITLE
`ultralytics 8.3.48` split PyPI publish jobs for security

### LINKED ISSUE #18027: Discrepancy between what's in GitHub and what's been published to PyPI for v8.3.41
# Bug
Code in the published wheel 8.3.41 is not what's in GitHub and appears to invoke mining. Users of ultralytics who install 8.3.41 will unknowingly execute an xmrig miner.

Examining the file utils/download.py, the contents in the published wheel are not representative of what's in GitHub:

```
$ wget -q https://files.pythonhosted.org/packages/d0/99/13d92174aa6a470d348a95e31164769f2cdf77838ea3c3e3fd476285777d/ultralytics-8.3.41-py3-none-any.whl
$ unzip -q ultralytics-8.3.41-py3-none-any.whl 
$ cat ultralytics/utils/downloads.py | sed -n "278,301p"
```

For the 8.3.41 tag, the content looks quite different:
https://github.com/ultralytics/ultralytics/blob/v8.3.41/ultralytics/utils/downloads.py#L278

# Environment
Ultralytics 8.3.41

# Minimal Reproducible Example

```
$ wget -q https://files.pythonhosted.org/packages/d0/99/13d92174aa6a470d348a95e31164769f2cdf77838ea3c3e3fd476285777d/ultralytics-8.3.41-py3-none-any.whl
$ unzip -q ultralytics-8.3.41-py3-none-any.whl 
$ cat ultralytics/utils/downloads.py | sed -n "278,301p"
```

### PR DESCRIPTION
**Pull request by glenn-jocher**
Per recommendation by @webknjaz in https://github.com/ultralytics/ultralytics/issues/18027#issuecomment-2525340727

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
This PR improves and restructures the GitHub Actions workflow for publishing the Ultralytics package to PyPI, introducing new job stages and better automation for version handling, notifications, and error management. 🔄🚀  

### 📊 Key Changes  
- **Refined Workflow Structure:**  
  - Split the workflow into four distinct stages: `check`, `build`, `publish`, and `notify`, for enhanced clarity and task separation.  
- **Version Check Improvements:**  
  - Added functionality to determine new version increments (`current_tag`, `previous_tag`) and decide if publishing is needed.  
- **Artifact Upload/Download:**  
  - Introduced artifact handling to share build outputs (`dist`) across stages.  
- **Improved Notifications:**  
  - Enhanced success/failure Slack notifications for better communication during releases.  
- **Bumped Package Version:**  
  - Incremented `__version__` to `8.3.48`.  

### 🎯 Purpose & Impact  
- **Purpose:**  
  - Simplify the CI/CD publishing process for PyPI by breaking it into modular, reusable stages.  
  - Improve automation in versioning and release handling.  
- **Impact:**  
  - Streamlines future package releases with less manual intervention and clearer logging.  
  - Reduces errors and allows for better debugging during failed workflows.  
  - Keeps collaborators promptly informed about successful or unsuccessful actions with detailed Slack messages. 💬✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your PR, including what problem it solves or what feature it adds. If applicable, please reference any [relevant issues](https://github.com/ultralytics/ultralytics/issues) or discussions in the description.
- ✅ **Synchronize with Source**: Make sure your branch is up-to-date with the latest `main` branch from `ultralytics/ultralytics`. You can do this by clicking the 'Update branch' button, or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Confirm that all [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If they’re failing, you may need to address the issues in your code before this PR can be reviewed.
- ✅ **Update Documentation**: If this PR introduces new functionality or changes any API behavior, ensure the [documentation](https://docs.ultralytics.com) is updated to reflect these changes.
- ✅ **Add or Update Tests**: Include test cases to cover any new functionality or modifications, and verify that all existing tests pass successfully.
- ✅ **Sign the CLA**: If this is your first contribution to Ultralytics, please sign the [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by commenting "I have read the CLA Document and I sign the CLA" in this PR.
- ✅ **Keep Changes Minimal**: Wherever possible, minimize the size of changes to focus only on the specific improvements or fixes. This practice enhances code clarity and simplifies review. Or as Bruce Lee famously said, _“Hack away the unessential.”_ 🥋

For more detailed guidance, check out our [Contributing Guide](https://docs.ultralytics.com/help/contributing).

This is an automated response to help streamline the PR review process. An Ultralytics engineer will follow up to review and address your PR shortly. 🚦  

Thank you for helping to improve Ultralytics! 🚀

**Review by pderrenger — APPROVED** [review]

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -13,32 +13,23 @@ on:
...
 jobs:
-  publish:
+  check:
     if: github.repository == 'ultralytics/ultralytics' && github.actor == 'glenn-jocher'
-    name: Publish
     runs-on: ubuntu-latest
-    environment: # for GitHub Deployments tab
-      name: Release - PyPI
-      url: https://pypi.org/p/ultralytics
     permissions:
```
Comment (webknjaz) [inline .github/workflows/publish.yml:19]: This sets up the job permissions. Some people additionally set the workflow-level permissions to `none` to lower the impact for any new jobs might have. Consider doing so as well.

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -49,30 +40,63 @@ jobs:
...
+    steps:
+      - uses: actions/download-artifact@v4
+        with:
+          name: dist
+          path: dist/
+      - uses: pypa/gh-action-pypi-publish@release/v1
+
+  notify:
+    needs: [check, publish]
+    if: always() && needs.check.outputs.increment == 'True'
```
Comment (akx) [inline .github/workflows/publish.yml:96]: `if: always() && ...`? 😄

Comment (webknjaz) [reply .github/workflows/publish.yml:96]: Yes, it's necessary to overcome certain corner cases.

**Pull request merged**